crates/grafbase-local-backend/src/dev/hot_reload.rs

```diff
@@ -1,18 +1,253 @@
-use super::configurations::get_and_merge_configurations;
+use super::configurations::{get_and_merge_configurations, DevConfiguration};
 use super::subgraphs::{get_subgraph_sdls, SubgraphCache};
-use super::FullGraphRef;
+use crate::dev::subgraphs::CachedIntrospectedSubgraph;
 use crate::errors::BackendError;
-use futures::lock::Mutex;
 use gateway_config::Config;
-use notify::RecursiveMode;
-use notify_debouncer_full::{new_debouncer, DebounceEventResult};
+use grafbase_graphql_introspection::introspect;
+use notify::{RecommendedWatcher, RecursiveMode};
+use notify_debouncer_full::{new_debouncer, DebounceEventResult, Debouncer, RecommendedCache};
+use std::collections::HashSet;
+use std::sync::Arc;
 use std::{path::PathBuf, time::Duration};
 use tokio::runtime::Handle;
 use tokio::sync::broadcast::Receiver;
 use tokio::sync::mpsc;
+use tokio::time::MissedTickBehavior;
 use tokio_stream::{wrappers::ReceiverStream, StreamExt};
+use tokio_util::sync::CancellationToken;
 
-const WATCHER_DEBOUNCE_TIME: Duration = Duration::from_secs(2);
+struct SubgraphWatcher {
+    watcher: Option<Debouncer<RecommendedWatcher, RecommendedCache>>,
+    cancellation_token: Option<CancellationToken>,
+}
+
+impl SubgraphWatcher {
+    fn new() -> Self {
+        Self {
+            watcher: None,
+            cancellation_token: None,
+        }
+    }
+
+    fn stop(&mut self) {
+        self.watcher = None;
+        if let Some(ref poller_cancellation_token) = self.cancellation_token {
+            poller_cancellation_token.cancel();
+        }
+        self.cancellation_token = None;
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    async fn start(
+        &mut self,
+        sender: mpsc::Sender<(String, Arc<Config>)>,
+        subgraph_cache: &'static SubgraphCache,
+        overridden_subgraphs: Arc<HashSet<String>>,
+        merged_configuration: Arc<Config>,
+        graph_overrides_path: Option<&'static PathBuf>,
+    ) -> Result<(), BackendError> {
+        // skip if there's no local subgraphs
+        if overridden_subgraphs.is_empty() {
+            return Ok(());
+        }
+
+        self.cancellation_token = Some(CancellationToken::new());
+
+        self.introspection_poller(
+            sender.clone(),
+            subgraph_cache,
+            overridden_subgraphs.clone(),
+            merged_configuration.clone(),
+            graph_overrides_path,
+        )
+        .await?;
+
+        self.schema_file_watcher(
+            sender,
+            subgraph_cache,
+            overridden_subgraphs,
+            merged_configuration,
+            graph_overrides_path,
+        )
+        .await
+    }
+
+    async fn introspection_poller(
```

Comment:
Does this function need to be async?

Also might be clearer if it was named `spawn_introspection_poller` - hard to tell at the callsite whether it spawns something or actually does the polling.

Reply:
RE the first point - WDYM? It calls multiple async functions and does the polling in parallel

Reply:
renamed

Reply:
Done